synapse/storage/databases/state/store.py

```diff
@@ -228,6 +247,144 @@ def _get_state_for_group_using_cache(
+
+        async def _the_request() -> StateMap[str]:
+            group_to_state_dict = await self._get_state_groups_from_groups(
+                (group,), state_filter=db_state_filter
+            )
+
+            # Now let's update the caches
+            self._insert_into_cache(
+                group_to_state_dict,
+                db_state_filter,
+                cache_seq_num_members=cache_sequence_m,
+                cache_seq_num_non_members=cache_sequence_nm,
+            )
+
+            # Remove ourselves from the in-flight cache
+            group_request_dict = self._state_group_inflight_requests[group]
+            del group_request_dict[db_state_filter]
+            if not group_request_dict:
+                # If there are no more requests in-flight for this group,
+                # clean up the cache by removing the empty dictionary
+                del self._state_group_inflight_requests[group]
+
+            return group_to_state_dict[group]
+
+        # We don't immediately await the result, so must use run_in_background
+        # But we DO await the result before the current log context (request)
+        # finishes, so don't need to run it as a background process.
+        request_deferred = run_in_background(_the_request)
+        observable_deferred = ObservableDeferred(request_deferred)
+
+        # Insert the ObservableDeferred into the cache
+        group_request_dict = self._state_group_inflight_requests.setdefault(group, {})
+        group_request_dict[db_state_filter] = observable_deferred
+
+        return await make_deferred_yieldable(observable_deferred.observe())
+
+    async def _get_state_for_group_using_inflight_cache(
+        self, group: int, state_filter: StateFilter
+    ) -> MutableStateMap[str]:
+        """
+        Gets the state at a state group, potentially filtering by type and/or
+        state key.
+
+        Args:
+            group: ID of the state group for which we want to get state
+            state_filter: the state filter used to fetch state from the database
+        Returns:
+            state map
+        """
+
+        # first, figure out whether we can re-use any in-flight requests
+        # (and if so, what would be left over)
+        (
+            reusable_requests,
+            state_filter_left_over,
+        ) = self._get_state_for_group_gather_inflight_requests(group, state_filter)
+
+        if state_filter_left_over != StateFilter.none():
+            # Fetch remaining state
+            remaining = await self._get_state_for_group_fire_request(
+                group, state_filter_left_over
+            )
+            assembled_state: MutableStateMap[str] = dict(remaining)
+        else:
+            assembled_state = {}
+
+        gathered = await make_deferred_yieldable(
+            defer.gatherResults(
+                (r.observe() for r in reusable_requests), consumeErrors=True
+            )
+        ).addErrback(unwrapFirstError)
```

Comment:
Does it make sense to both consume errors and unwrap the first one?

Reply:
Your question sounds like you think it's not. I'm struggling to understand the docstrings, but it looks like:

* `gatherResults` constructs a `DeferredList` with `fireOnOneErrback = True`
* ```
            @param fireOnOneErrback: (keyword param) a flag indicating that this
            L{DeferredList} will fire when the first L{Deferred} in
            C{deferredList} fires with a failure result without waiting for any
            of the other Deferreds.  When this flag is set, if a Deferred in the
            list errbacks, the DeferredList will errback with a L{FirstError}
            failure wrapping the failure of that Deferred.
    ```
     * i.e. I think this will give you a `FirstError(Exception)` rather than `Exception` if anything happens
* `unwrapFirstError` looks like this:
     ```python
     def unwrapFirstError(failure: Failure) -> Failure:
         # defer.gatherResults and DeferredLists wrap failures.
         failure.trap(defer.FirstError)
         return failure.value.subFailure  # type: ignore[union-attr]  # Issue in Twisted's annotations
     ```
    * essentially `.mapError(lambda err: err.value.subFailure)` to my eye? `self.trap(x)` seems to be more or less `if not isinstance(self, x): raise self` (it wasn't obvious to me)

I don't seem to be the first using it this way: https://github.com/matrix-org/synapse/blob/7468723697e4d292315ce807b5000062a02b37be/synapse/util/caches/descriptors.py#L474-L476 .

Does this sound right now or is there something missing?

Reply:
Yeah, that sounds reasonable. Thanks!